src/unix/io_uring.c

```diff
@@ -0,0 +1,1340 @@
+                    ret = -EBADF;
+                }
+            }
+            if (!ret) {
+                fdesc old = iour->files[offset + i];
+                if (old)
+                    fdesc_put(old);
+                iour->files[offset + i] = f;
+            } else {
+                if (i > 0)
+                    ret = i;
+                break;
+            }
+        }
+    }
+    iour_unlock(iour);
+    if (!ret)
+        ret = count;
+    return ret;
+}
+
+static boolean iour_submit(io_uring iour, struct io_uring_sqe *sqe)
+{
+    iour_debug("opcode %d, flags 0x%x, user_data %ld", sqe->opcode, sqe->flags,
+        sqe->user_data);
+    fdesc f = 0;
+    s32 res;
+    if (sqe->flags & ~(IOSQE_FIXED_FILE | IOSQE_ASYNC)) {
+        /* non-supported flags */
+        res = -EINVAL;
+        goto complete;
+    }
+    switch(sqe->opcode) {
+    case IORING_OP_READV:
+    case IORING_OP_WRITEV:
+    case IORING_OP_READ_FIXED:
+    case IORING_OP_WRITE_FIXED:
+    case IORING_OP_POLL_ADD:
+    case IORING_OP_READ:
+    case IORING_OP_WRITE:
+        if (sqe->flags & IOSQE_FIXED_FILE) {
+            iour_lock(iour);
+            int fd = sqe->fd;
+            if ((fd >= 0) && (fd < iour->file_count)) {
+                f = iour->files[fd];
+                if (f)
+                    f->refcnt++;
+            }
+            iour_unlock(iour);
+        } else
+            f = fdesc_get(current->p, sqe->fd);
+        if (!f) {
+            res = -EBADF;
+            goto complete;
+        }
+        break;
+    default:
+        break;
+    }
+    switch (sqe->opcode) {
+    case IORING_OP_NOP:
+        res = 0;
+        goto complete;
+    case IORING_OP_READV:
+    case IORING_OP_WRITEV: {
+        if (sqe->buf_index) {
+            res = -EINVAL;
+            goto complete;
+        }
+        struct iovec *iov = pointer_from_u64(sqe->addr);
+        u32 len = sqe->len;
+        boolean write = (sqe->opcode == IORING_OP_WRITEV);
+        if (!validate_iovec(iov, len, write)) {
+            res = -EFAULT;
+            goto complete;
+        }
+        iour_iov(iour, f, write, iov, len, sqe->off, sqe->user_data);
+        break;
+    }
+    case IORING_OP_READ_FIXED:
+    case IORING_OP_WRITE_FIXED:
+        res = 0;
+        iour_lock(iour);
+        u16 buf_index = sqe->buf_index;
+        if (buf_index >= iour->buf_count) {
+            if (iour->buf_count != 0)
+                res = -EINVAL;
+            else
+                res = -EFAULT;
+        } else {
+            struct iovec *iov = &iour->bufs[buf_index];
+            void *buf = pointer_from_u64(sqe->addr);
+            u32 len = sqe->len;
+            boolean write = sqe->opcode == IORING_OP_WRITE_FIXED;
+
+            /* If we have the capability to pin memory pages, as done in Linux
+             * with pin_user_pages(), we will validate the memory at buffer
+             * registration time and avoid validating it here. */
```

Comment:
If I understood it correctly, page pinning in general prevents a given memory range from being deallocated for a given process. If we don't have that, an application could register a set of buffers with io_uring, then free (unmap) that memory, then request I/O on that memory, at which point a page fault would happen (in kernel mode) and the relevant address wouldn't be found in the process vmap, i.e. the kernel wouldn't be able to handle the fault.

Reply:
I think the pinning is to take a reference for any physical pages that are behind the mapping at the time of the call, but to my knowledge it doesn't guarantee anything about the v->p mapping of those pages or whether a virtual address range remains mapped or consistent. But maybe it does prevent those pages from being drained from the page cache or swapped out (provided the process isn't doing something to remove a mapping). I'm not clear on that detail.

But it may still be relevant here for any calls which do zero-copy (or less-copy?) directly from user buffers - those indeed will need to have pages pinned. Our equivalent - at this time anyway - is to add a reference to the pagecache_page so that the (physical) page cannot be drained from the cache or otherwise freed. (We maintain those references in the cache, for we don't have anything like the struct page array in Linux.) But that still won't prevent the process from unmapping pages or otherwise doing something that might cause a fault on translation.

Reply:
OK, got it, I changed the code so that it does iovec validation when buffers are registered, instead of validating the memory at each READ_FIXED/WRITE_FIXED operation.
The io_uring_register man page in the IORING_REGISTER_BUFFERS description says "The buffers associated with the iovecs will be locked in memory", so it looks like Linux does prevent swapping out relevant pages. But I don't see any I/O ops (among those that can be triggered via io_uring) that _require_ pages to be pinned, so I guess we can leave it as it is for now.